### Yusif Hajizade CS-020 id: 22022735
# **Artificial Intelligence Practical Work 4**
## **Lab : Artificial Neural Network training, from scratch**
#### **Intoduction:**
In this practical work, we aim to build an artificial neural network (ANN) from scratch and train it using the Iris dataset. The ANN will be implemented using Python, and we'll utilize matrices to represent both the parameters of the model and the dataset.

#### **Objectives and Goals:**
Our objective during this lab is to build an artificial neural network from scratch and train it on the Iris
dataet.

**Specific objectives:**
- Observe the data you want to learn from, understand their nature and how to adapt them (if
needed) to use them with a neural network model.
- Understand how a neural network model works so as to fully implement one in a program.


### **1. Network Architecture and Data**

In this practical work, we aim to build an artificial neural network (ANN) from scratch and train it using the Iris dataset. The ANN will be implemented using Python, and we'll utilize matrices to represent both the parameters of the model and the dataset.


In [12]:
# import all necessary libraries 
import pandas as pd
import numpy as np
from sklearn.utils import shuffle



#### **Loading the data into a matrix:**

The data file is read using a method in the class
DataLib and used in the main. The data matrix has type float, 2 dimentsions and it is used to instanciate
the neural network. You may also have noticed that the data are shuffled before they are returned as
a 2D array

In [11]:
class DataLib:
    @staticmethod
    def load_data(file_path):
        # Load the data from the CSV file into a DataFrame
        data = pd.read_csv(file_path)
        
        # Drop the "Id" column
        data.drop("Id", axis=1, inplace=True)
        
        # Shuffle the data
        data = data.sample(frac=1).reset_index(drop=True)
        
        # Extract the values as a matrix
        data_matrix = data.values
        
        return data_matrix


# Example usage
file_path = "Iris.csv"
iris_data = DataLib.load_data(file_path)
print(iris_data)

[[5.4 3.0 4.5 1.5 'Iris-versicolor']
 [6.4 3.2 5.3 2.3 'Iris-virginica']
 [4.3 3.0 1.1 0.1 'Iris-setosa']
 [5.8 2.7 5.1 1.9 'Iris-virginica']
 [5.5 2.4 3.7 1.0 'Iris-versicolor']
 [4.6 3.1 1.5 0.2 'Iris-setosa']
 [5.0 3.3 1.4 0.2 'Iris-setosa']
 [5.6 2.7 4.2 1.3 'Iris-versicolor']
 [6.7 3.3 5.7 2.5 'Iris-virginica']
 [6.4 2.7 5.3 1.9 'Iris-virginica']
 [5.0 3.2 1.2 0.2 'Iris-setosa']
 [5.4 3.9 1.7 0.4 'Iris-setosa']
 [5.9 3.0 5.1 1.8 'Iris-virginica']
 [4.9 2.5 4.5 1.7 'Iris-virginica']
 [6.7 3.0 5.2 2.3 'Iris-virginica']
 [4.4 2.9 1.4 0.2 'Iris-setosa']
 [5.2 3.5 1.5 0.2 'Iris-setosa']
 [6.0 2.2 4.0 1.0 'Iris-versicolor']
 [5.1 3.3 1.7 0.5 'Iris-setosa']
 [6.7 3.0 5.0 1.7 'Iris-versicolor']
 [7.7 2.6 6.9 2.3 'Iris-virginica']
 [5.2 4.1 1.5 0.1 'Iris-setosa']
 [5.0 2.3 3.3 1.0 'Iris-versicolor']
 [5.3 3.7 1.5 0.2 'Iris-setosa']
 [7.3 2.9 6.3 1.8 'Iris-virginica']
 [4.9 3.1 1.5 0.1 'Iris-setosa']
 [6.3 2.5 4.9 1.5 'Iris-versicolor']
 [7.2 3.0 5.8 1.6 'Iris-virginica']
 [4.9 3.0 1.4 0.2 

In [ ]:
import numpy as np

class NNLib:
    @staticmethod
    def initialize_weights(input_size, hidden_size, output_size):
        """
        Initialize weights for the neural network.

        Parameters:
        input_size (int): Size of the input layer.
        hidden_size (int): Size of the hidden layer.
        output_size (int): Size of the output layer.

        Returns:
        tuple: Initialized weights for the input-hidden and hidden-output layers.
        """
        # Initialize weights for input-hidden layer
        W1 = np.random.randn(input_size, hidden_size)
        b1 = np.zeros((1, hidden_size))

        # Initialize weights for hidden-output layer
        W2 = np.random.randn(hidden_size, output_size)
        b2 = np.zeros((1, output_size))

        return W1, b1, W2, b2

    @staticmethod
    def sigmoid(x):
        """
        Sigmoid activation function.

        Parameters:
        x (numpy.ndarray): Input to the sigmoid function.

        Returns:
        numpy.ndarray: Output of the sigmoid function.
        """
        return 1 / (1 + np.exp(-x))

    @staticmethod
    def sigmoid_derivative(x):
        """
        Derivative of the sigmoid activation function.

        Parameters:
        x (numpy.ndarray): Input to the sigmoid function.

        Returns:
        numpy.ndarray: Derivative of the sigmoid function.
        """
        return x * (1 - x)

    @staticmethod
    def softmax(x):
        """
        Softmax activation function.

        Parameters:
        x (numpy.ndarray): Input to the softmax function.

        Returns:
        numpy.ndarray: Output of the softmax function.
        """
        exp_values = np.exp(x - np.max(x, axis=1, keepdims=True))
        return exp_values / np.sum(exp_values, axis=1, keepdims=True)

    @staticmethod
    def softmax_derivative(x):
        """
        Derivative of the softmax activation function.

        Parameters:
        x (numpy.ndarray): Input to the softmax function.

        Returns:
        numpy.ndarray: Derivative of the softmax function.
        """
        # The derivative of the softmax function is calculated during backpropagation
        # as part of the cross-entropy loss gradient computation.
        # Hence, we do not explicitly need to implement it here.
        pass

    @staticmethod
    def cross_entropy_loss(y_true, y_pred):
        """
        Cross-entropy loss function.

        Parameters:
        y_true (numpy.ndarray): True labels.
        y_pred (numpy.ndarray): Predicted probabilities.

        Returns:
        float: Cross-entropy loss.
        """
        m = y_true.shape[0]
        log_likelihood = -np.log(y_pred[range(m), y_true.argmax(axis=1)])
        loss = np.sum(log_likelihood) / m
        return loss



#### **Question 1.1: Dealing with Matrices**
1. **How are the weights initialized?**
   - The weights are initialized randomly using a Gaussian distribution with mean 0 and standard deviation 1.
   
2. **Which activation functions are available in NNLib.py?**
   - The available activation functions include the sigmoid function, tanh function, and ReLU (Rectified Linear Unit) function.
   
3. **Since it's a classification problem, what activation function will be used for the output layer?**
   - For the output layer of a classification problem, typically the softmax function is used to produce probabilities for each class.

4. **What cost function is available?**
   - The available cost function is the cross-entropy loss function, commonly used for classification problems.

#### **Question 1.2: Neural Network Architecture**
1. **What proportion of the data is held for training and testing?**
   - Typically, the data is split into 80% for training and 20% for testing.

2. **What is contained in X_train, X_test, Y_train, and Y_test?**
   - X_train and X_test contain the features (input data) for training and testing, respectively, while Y_train and Y_test contain the corresponding labels (output data).

3. **How many hidden layers are used, and how many units are there per layer?**
   - The number of hidden layers and units per layer depends on the specific architecture chosen for the neural network.


### **2. Preparing the data**